### ISSUE TITLE
Switch statement when used with more then 6 cases will generate a compile error

### ISSUE THREAD

**Issue by aronmek** [nikhilk/scriptsharp#56]
when I use a switch with a static condition for Ex.

```
public class Class1
{

    public Class1()
    {
        switch ("123")
        {
            case "1":
                break;
            case "2":
                break;
            case "3":
                break;
            case "4":
                break;
            case "5":
                break;
            case "6":
                break;
            case "7":
                break;
        }
    }
}
```

I don't have any problems but when I use it with a variable condition for Ex.

```
public class Class1
{

    public Class1()
    {
        string test = "123";
        switch (test)
        {
            case "1":
                break;
            case "2":
                break;
            case "3":
                break;
            case "4":
                break;
            case "5":
                break;
            case "6":
                break;
            case "7":
                break;
            case "8":
                break;
        }
    }
}
```

the compiler crashes.

My script sharp version is 0.7.0.0 and I'm using VWD 2010

**Comment by nikhilk** [comment]
What you are seeing is the c# compiler crashing. The script# compiler handles this fine. Not sure why it is happening, but it seems there is something about the custom mscorlib causing this. It might be because c# starts emitting code (eg string interning) when you have more than 6 strings in a switch.

Try using an enum marked with [NamedValues] - it gives better c# semantics and code experience like intelligence + typo protection and translates to string literals in generated script.

**Comment by aronmek** [comment]
I never knew you can give named values for enums in c#, How can I do that?

**Comment by nikhilk** [comment]
[NamedValues]
public enum Values {
  X = 1,
  Y = 2
}

The script will have "x", and "y" rather than Values.X and Values.Y.

**Comment by aronmek** [comment]
Got it, I'll try it tomorrow because I'm not quite sure what the Enum.Parse() will return me 

That's the code

<pre>
PanelHeaderIds CurrentHeaderId = (PanelHeaderIds)Enum.Parse(typeof(PanelHeaderIds),Panel.GetAttribute("id"));
            switch (CurrentHeaderId)
            {
                case PanelHeaderIds.Dispatch:
                    ReturnValue = Run.Pcr.Dispatch;
                    break;
                case PanelHeaderIds.Demographic:
                    ReturnValue = Run.Pcr.Demographic;
                    break;
                case PanelHeaderIds.Assessment:
                    ReturnValue = Run.Pcr.Assessment;
                    break;
                case PanelHeaderIds.VitalSigns:
                    ReturnValue = Run.Pcr.Vitals;
                    break;
                case PanelHeaderIds.Treatment:
                    ReturnValue = Run.Pcr.Treatment;
                    break;
                case PanelHeaderIds.Disposition:
                    ReturnValue = Run.Pcr.Disposition;
                    break;
                case PanelHeaderIds.Protocols:
                    ReturnValue = Run.Pcr.Protocols;
                    break;
            }
</pre>

And this is the enum

<pre>
[NamedValues]
    public enum PanelHeaderIds
    {
        [ScriptName("panel1Header")]
        Dispatch = 0,
        [ScriptName("panel2Header")]
        Demographic = 1,
        [ScriptName("panel3Header")]
        Assessment = 2,
        [ScriptName("panel4Header")]
        VitalSigns = 3,
        [ScriptName("panel5Header")]
        Treatment = 4,
        [ScriptName("panel6Header")]
        Disposition = 5,
        [ScriptName("panel7Header")]
        Protocols = 6
    }
</pre>

**Comment by nikhilk** [comment]
You don't have to use Enum.Parse. You'd have the following in your code:

```
switch ((PanelHeaderIds)Panel.GetAttribute("id")) {
    ...
}
```

**Comment by aronmek** [comment]
I had to do it like this 
<code>
switch ((PanelHeaderIds)(object)Panel.GetAttribute("id"))
</code>
at the end and it works. 
Thanks I really appreciate your time and your "Project" Keep it up.

**Comment by nikhilk** [comment]
Tracking the investigation results... the c# compiler needs the following members on the generic Dictionary class:
1. ctor with int parameters
2. void Add(TKey, TValue)
3. bool TryGetValue(TKey, out TValue)

In the absense of these, the c# compiler crashes.

These APIs are not supportable in script yet (Eg. out params not supported) ... so it would be odd to add them for getting this switch case working, but not support app code calling into them.

**Comment by robitar** [comment]
Is there a way to decorate the members with the Editor Browsable attribute so they don't show up in intellisense?

**Comment by jimmygilles** [comment]
I have the same problem using switch.

# Here is an example:

``` C#
using System.Html;

namespace ScriptLibrary1
{
    public static class Class1
    {
        public static Boolean Test(String aValue)
        {
            switch (aValue)
            {
                case "val1": return true;
                case "val2": return true;
                case "val3": return true;
                case "val4": return true;
                case "val5": return true;
                case "val6": return true;
                case "val7": return true;
                case "val8": return true;
                default: return false;
            }
        }
    }
}
```

# And here is the result on VisualStudio (it's in french)

Erreur  7   Erreur interne du compilateur (0xc0000005 à l'adresse 56365E79) : 'CODEGEN' en est probablement la cause.

Une erreur interne s'est produite dans le compilateur. Pour résoudre ce problème, essayez de simplifier ou de changer le programme près des emplacements indiqués ci-dessous. Les emplacements situés au début de la liste sont les plus proches de l'erreur interne. Les erreurs telles que celle-ci peuvent être signalées à Microsoft à l'aide de l'option /errorreport.
    ScriptLibrary1
Erreur  1   Erreur interne du compilateur : étape 'BEGIN' ScriptLibrary1
Erreur  6   Erreur interne du compilateur : étape 'CODEGEN'   C:\Users\Jimmy Gilles\Programmation\ScriptLibrary1\ScriptLibrary1\Class1.cs ScriptLibrary1
Erreur  13  Erreur interne du compilateur : étape 'CODEGEN'   C:\Users\Jimmy Gilles\Programmation\ScriptLibrary1\ScriptLibrary1\Class1.cs 11  13  ScriptLibrary1
Erreur  12  Erreur interne du compilateur : étape 'CODEGEN' symbole 'ScriptLibrary1.Class1.Test(string)'  C:\Users\Jimmy Gilles\Programmation\ScriptLibrary1\ScriptLibrary1\Class1.cs 9   31  ScriptLibrary1
Erreur  3   Erreur interne du compilateur : étape 'COMPILE'   ScriptLibrary1
Erreur  4   Erreur interne du compilateur : étape 'COMPILE'   C:\Users\Jimmy Gilles\Programmation\ScriptLibrary1\ScriptLibrary1\Class1.cs ScriptLibrary1
Erreur  5   Erreur interne du compilateur : étape 'COMPILE' symbole '<espace de noms global>' C:\Users\Jimmy Gilles\Programmation\ScriptLibrary1\ScriptLibrary1\Class1.cs ScriptLibrary1
Erreur  8   Erreur interne du compilateur : étape 'COMPILE' symbole 'ScriptLibrary1'  C:\Users\Jimmy Gilles\Programmation\ScriptLibrary1\ScriptLibrary1\Class1.cs 5   11  ScriptLibrary1
Erreur  9   Erreur interne du compilateur : étape 'COMPILE' symbole 'ScriptLibrary1.Class1'   C:\Users\Jimmy Gilles\Programmation\ScriptLibrary1\ScriptLibrary1\Class1.cs 7   25  ScriptLibrary1
Erreur  10  Erreur interne du compilateur : étape 'COMPILE' symbole 'ScriptLibrary1.Class1.Test(string)'  C:\Users\Jimmy Gilles\Programmation\ScriptLibrary1\ScriptLibrary1\Class1.cs 9   31  ScriptLibrary1
Erreur  11  Erreur interne du compilateur : étape 'COMPILE' symbole 'ScriptLibrary1.Class1.Test(string)'  C:\Users\Jimmy Gilles\Programmation\ScriptLibrary1\ScriptLibrary1\Class1.cs 9   31  ScriptLibrary1
Erreur  2   Erreur interne du compilateur : étape 'EMIT'  ScriptLibrary1

**Comment by nikhilk** [comment]
Jimmy, the problem is about the switch, and how the c# compiler uses dictionaries to handle switch statements with 6+ case statements, so not sure how your problem is any different than the one originally reported.

**Comment by nikhilk** [comment]
Fix included in 0.7.4.0 - APIs required by csc.exe have been added, but marked as obsolete (with error) to prevent them being used, as they aren't supported at runtime.
